### ISSUE TITLE
Keeping track of installed repos.

### ISSUE THREAD

**Issue by gabyx** [rycus86/githooks#55]
Scenario I am not sure about:
- RepoA (not a **single** install) triggers an update (by `git commit`) not from the console  but over a gui (no /dev/tty:
     - Uses `sh <install.sh` which launches a **global** update 👍 :
        - Update the templates and cli
        - Install into existing repos by the search dir if existing, otherwise do nothing

- Inside RepoA (not a **single** install), `git hooks update`
   - Same procedure as above...

The install into RepoA is (I think) not ensured which is a bit bad.

Wouldn't it make sense, to somehow (instead or additional to a search dir) have a list of installed repos, we keep track of, such that when ever a global update/install happens, we can install the new hooks also into the bookkeeping list (`~/.githooks/previous`) (if they are still existing at the location).

Options:
- `githooks.previous.repositories = "repoA;repoB" ` : Bookkeeping  of installed repos. or into the file `~/.githooks/previous`

```ini
[githooks "previous"]
	searchdir = ~/.bash-git-prompt
        repositories = repoA;repoB
```
What do you think. That would make update behavior a bit less cumbersome, and error prone. (user having a wrong search dir and such. Of course we can always clean the bookkeeping file or variable of any non existing repos, such that its always accurate. I tend towards a file, which is nicer to read...

**Comment by gabyx** [comment]
I was thinking about that, because of update with critical bugfixes... (hopefully not much in the future)
A more reliable update procedure would be awesome :-). If I am mistaken about 
> The install into RepoA is (I think) not ensured which is a bit bad.
Do you know a test which tests this?

**Comment by gabyx** [comment]
Ok we cannot update our hooks when inside RepoA's `base_template.sh` (`git commit`). 
- Maybe install an Exit Trap which checks if an install is postponing and do it there (cumbersome)
- Other solution: Make an option `githooks.failIfNotCurrentVersion` which fails any hook if the current base_template.sh has a different version than `git hooks version`. 
That would resolve running non-updated hooks, leaving the update process to the user.

The suggestion with the tracked install dirs would be nice anyway.

**Comment by gabyx** [comment]
The following test fails and demonstrates the weird behavior:

```shell
#!/bin/sh
# Test:
#   Trigger update

if ! sh /var/lib/githooks/install.sh; then
    echo "! Failed to execute the install script"
    exit 1
fi

git hooks config set search-dir /tmp/test101 || exit 1

mkdir -p /tmp/test101 &&
    cd /tmp/test101 &&
    git init || exit 1

# Make hook expired
sed -i 's/^# Version: .*/# Version: 0/' .git/hooks/* || exit 1
out=$(git commit --allow-empty -m "Test" 2>&1)

if ! echo "$out" | grep -q "There is a new Githooks update available"; then
    echo "Expected update to happen"
    exit 1
fi

if ! grep -q "# Version 0" .git/hooks/commit-msg; then
    echo "This running hook should not have been updated during execution!!"
    exit 1
fi
```

**Comment by gabyx** [comment]
Or is it like that since the first command to execute is at the bottom of the file, that the whole script gets read anyway, so its safe to overwrite the hook during update...?

**Comment by gabyx** [comment]
See the small easy fix #56 : Which at least for UNIX fixes the problem.

**Comment by gabyx** [comment]
I tried keeping track of all installed repos. git init cannot be hijacked somehow leaving this endeavor quite cumbersome and stupid. However: the following needs discussion:
- non single installed hooks can triggers a global install which only installs into existing by the search dir. the repo which initiated this might not be touched.
- for non single installed repos: in any hook: we should not only check for an update on  the server but also after that if the templates in the install folder are newer and instantly replace them. In this way we make sure any existing repo somewhere on the disk gets the latest stuff.

This doesnt seem to be a major change and seems feasible. Of course we still have the option coreHoolsPath (if its appropriate to use) which circumvents all of these troubles. I will try a PR. Thx for the input.

**Comment by gabyx** [comment]
I think I found a nice way of keeping track of repos using githooks hooks without performance issues in all cases. :)

**Comment by rycus86** [comment]
I'll need to check the PR properly as it ended up a bit big. :)
From a quick look, there are some calls there that are not POSIX and would fail on OS X for example, like `realpath` or `sed -i`.
I thought shellcheck would catch these, if not then I'll have to look into that.

What's the problem you're trying to solve here? It's not exactly clear, sorry.
Is this about an update overwriting the running hook? (which you also tackled with a smaller PR)

In general, the less state we have the better, so I'd like to understand first why you need this.

Thanks!

**Comment by gabyx** [comment]
Ou, sad, ok shellcheck didnt catch these... 
Ehm the PR needs some more work, I need to make the changes tonight as I have not pushed th stuff.

The thing about the global update mechanism is a bit confuse so far, thats why this PR.
I tried to describe it above: The install into a set search-dir is applied from a "update-triggering" hook which is running. And this procedure is a bit vague. Since the `install_into_existing_repo...` can end up installing the hooks also into `single` install repos (as you probably dit not intend) and its also too agnostic. I would have rather clear behavior what is done in the background: Store the installed repos we keep track of (only non-single installs) and then when some repo triggers an update , do the update over these registered repos. The install with a prompt asking for a search-dir and installing into the ones found is only done when really doing an install **manually**, which is perfectly fine, but not when an **update** is triggered by a hook, for that we use the register mechanism...

Thats much easier and clearer behavior. I try to find a solution without the realpath. I think its possible. It almost seems to me, with all these shell problems, shell is the wrong tool for such a tool with all these (nice) features...

**Comment by gabyx** [comment]
Regarding, the Bugfixes, should be easily mergable, and only contain some fixes apart from this PR :-).
Ok the thing with the search-dir is probably not mergable because of `real-path`

**Comment by gabyx** [comment]
Also a `Githooks-update-triggering` hook in `repoA/.git/hook/commit-msg` (for example) which is running, starts `install.sh` which does **not** update directly into `repoA` only by the search-dir... this is confuse...

**Comment by gabyx** [comment]
So for a better picture:

- single install repos only update them self, they do nothing else in the background.
- non-single install repos are all glued together basically, if one triggers an update (`install.sh`), all others also get the update too. Ensured by the registering mechanism (checked if the flag `githooks.autoupdate.registered` is unset in `base-template.sh`, if not its common-git-dir is set into the global list.

**Comment by rycus86** [comment]
Right. So you'd want single installs to also update everything, right?
As for the second part (non-single installs), os there anything wrong with the updates there specifically?

**Comment by gabyx** [comment]
Ok, you probably misread:
> So you'd want single installs to also update everything, right? (no, single install as explained in my first point, single installs should not have side effects, they care about them self and only them selfe, thats actually what I could get out from the code, and so far it looked as thats the behavior you wanted. (so far there was no problem with them)

Second point: non-single installs can have side effects (to single installs) depending on the set search-dir when a hook in such a non-single repo triggers and update. Furthermore and utmost: Since we only install to any repo which we find in the search-dir -> any `git clone` else where on the system does not get the update, and thats exactly my critical point, not even the repo of the update-initiating hook might get the update, because its not in the search-dir -> so the search-dir is a bit to wobbly IMO. Therefore, have a look at the changes I made, (it still needs tests of course) but the changes are not much, and they just manage a stupid list (even if the paths in the list migh get old/non-existing, dont worry its all safe).

**Comment by rycus86** [comment]
OK, thanks, that clears it up.
I think the way that you propose sounds good to me, and should also speed up the update a bit, replacing the slow search with just going through the list. :+1:

**Comment by gabyx** [comment]
Thx, lets make it bombproof.

**Comment by gabyx** [comment]
Merged #59
